### ISSUE TITLE
Microsoft.Data.Sqlite keeps handle on file even after close, and pool clear

### ISSUE THREAD

**Issue by ericnev** [dotnet/efcore#35010]
## Microsoft.Data.Sqlite keeps handle on file even after close, and pool clear

Using Microsoft.Data.Sqlite along with SQLitePCLRaw.bundle_e_sqlcipher, opening a file that's not a valid database file (throws error 26), the driver keeps the handle on the file, even in ReadOnly mode, and Pooling off. 

(We want to try and perform an integrity check on our databases, and alert the user of any issues, and can't copy off the file as it stands right now, as it's in-use.)

SqliteConnectionInternal calls SqliteException.ThrowExceptionForRC(rc, _db) but the file handle doesn't look to be let go via something like  sqlite3_close_v2(sqlite3 db), even on Dispose of the internal connection on Close() in the catch of the Open()? (_db.Dispose();)

```C#

using System.Data;
using Microsoft.Data.Sqlite;

Console.WriteLine("Db test");

string dbPath = "C:\\tmp\\test.db";
if (File.Exists(dbPath)) File.Delete(dbPath);
File.WriteAllText(dbPath, "this is not a database file.");
string connectionString = $"data source={dbPath};Password=12345;Mode=ReadOnly;Pooling=False";
SqliteCommand? cmd = null;
using SqliteConnection connection = new SqliteConnection(connectionString);
try
{
    connection.Open(); // exception thrown here
    cmd = connection.CreateCommand();
    cmd.CommandType = CommandType.Text;
    cmd.CommandText = "select count(*) from sqlite_master";
    var numTables = cmd.ExecuteScalar();
    bool everythingOk = (Convert.ToInt32(numTables) > 0);
    Console.WriteLine($"everythingOk:{everythingOk}");
}
catch (Exception e)
{
    Console.WriteLine(e);
    if (e is SqliteException sqliteException)
    {
        if (sqliteException.SqliteErrorCode == 26) // not a db
        {
            if (connection.State == ConnectionState.Open) connection.Close();
            SqliteConnection.ClearPool(connection);

            File.Move(dbPath, dbPath + $"{DateTime.Now.Ticks}.bad"); // exception thrown here
        }
    }
}
finally
{
    if (cmd != null)
    {
        cmd.Connection = null;
        cmd.Dispose();
    }
    if (connection.State == ConnectionState.Open) connection.Close();
    SqliteConnection.ClearPool(connection);
}

```

### Include version information

Microsoft.Data.Sqlite version: 8.0.10
Target framework: .NET 8
Operating system: Win 11
SQLitePCLRaw.bundle_e_sqlcipher 2.1.10

**Comment by ericnev** [comment]
The source of the behavior seems to be that the exception gets thrown in SqliteConnectionFactory.Instance.GetConnection(this). The exception is bubbled up before innerConnection can get defined (and closed/handled properly), and outerConnection.Handle isn't available, so the user can't clear the handle manually?

**Comment by ericnev** [comment]
Unsurprisingly, the same issue occurs on attempting to connect to a db with the wrong password. 

Locally, I put a try/catch around 

` if (SQLitePCLExtensions.EncryptionSupported() != false)
                {
                    // NB: Forces decryption. Throws when the key is incorrect.
                    ExecuteNonQuery(
                        "SELECT COUNT(*) FROM sqlite_master;",
                        connectionOptions.DefaultTimeout);
                }`

so the constructor can finish, and I can handle the 26 appropriately, with a freed file handle. Feel free to let me know if there's a better way to handle this.

**Comment by AyrA** [comment]
I'm experiencing the same issue. Doesn't matters if the DB already exists or not. Simply opening and closing the connection is enough.

Tested, and same behavior with 8.0.10, 8.0.11, 9.0.0

Minimal code for replication:

```csharp
var dbFile = Path.Combine(Path.GetTempPath(), "handle_leak_test.db3");
using (var conn = new SqliteConnection($"Data Source={dbFile}"))
{
    conn.Open();
    conn.Close();
}
File.Delete(dbFile); //Throws System.IO.IOException
```

**Comment by AyrA** [comment]
After some experimentation, it looks like closing the handle manually provides a workaround

```csharp
var dbFile = Path.Combine(Path.GetTempPath(), "handle_leak_test.db3");
using (var conn = new SqliteConnection($"Data Source={dbFile}"))
{
    conn.Open();
    var handle = conn.Handle; //"conn.Close();" immediately sets this to "null"
    conn.Close();
    handle?.Close(); // "handle?.Dispose();" also works, as well as "handle?.manual_close_v2();"
}
File.Delete(dbFile); //No longer throws System.IO.IOException
```

**Comment by ericnev** [comment]
> After some experimentation, it looks like closing the handle manually provides a workaround
> 
> var dbFile = Path.Combine(Path.GetTempPath(), "handle_leak_test.db3");
> using (var conn = new SqliteConnection($"Data Source={dbFile}"))
> {
>     conn.Open();
>     var handle = conn.Handle; //"conn.Close();" immediately sets this to "null"
>     conn.Close();
>     handle?.Close(); // "handle?.Dispose();" also works, as well as "handle?.manual_close_v2();"
> }
> File.Delete(dbFile); //No longer throws System.IO.IOException

Does that work for you even if the file isn't a database? My recollection is that Handle doesn't get set unless the Open() succeeds?

**Comment by AyrA** [comment]
For me it's set when calling `.Open();`

For some reason it doesn't cares that the file is not a database. I've created a file containing 4 MB of random bytes and the open call succeeds, and sets the handle too. I can even create and prepare a statement. Only when I run `.ExecuteNonQuery()` on the statement do I get "Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 26: 'file is not a database'."

**Comment by ericnev** [comment]
> For some reason it doesn't cares that the file is not a database. I've created a file containing 4 MB of random bytes and the open call succeeds, and sets the handle too. I can even create and prepare a statement. Only when I run `.ExecuteNonQuery()` on the statement do I get "Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 26: 'file is not a database'."
> 
> By catching the exception I can still close the internal handle.
> 
> [Test]
> public void TestCorruptDb()
> {
>     var dbFile = Path.GetTempFileName();
>     TestContext.Out.WriteLine("File is {0}", dbFile);
>     var bytes = new byte[4096 * 1024]; // 4 MB
>     Random.Shared.NextBytes(bytes);
>     File.WriteAllBytes(dbFile, bytes);
>     SqliteConnection? c = null;
>     try
>     {
>         c = new SqliteConnection($"Data Source={dbFile}");
>         c.Open(); //Works
>         Assert.That(c.Handle, Is.Not.Null);
> 
>         using (var cmd = c.CreateCommand())
>         {
>             cmd.CommandText = "CREATE TABLE \"test\"(\"id\" INTEGER PRIMARY KEY UNIQUE NOT NULL) STRICT;";
>             cmd.ExecuteNonQuery(); //Error 26
>         }
> 
>         c.Handle?.Close(); //Not reached
>         c.Close();
>     }
>     catch (Exception ex)
>     {
>         c?.Handle?.Close(); //Works, handle still exists
>         TestContext.Out.WriteLine(ex);
>         Assert.Fail(ex.Message);
>     }
>     finally
>     {
>         File.Delete(dbFile); //Works, handle is closed
>     }
> }

I see why, it's been a few weeks since I looked at it. If your provider doesn't support encryption (I'm using SqlCipher) you won't fall into the block that's causing the issue. If encryption is supported, the Open() call makes its own ExecuteNonQuery call that locks up the file, and never lets it go. Thanks for checking!

**Comment by ericnev** [comment]
> > No one had this problem since 2021?
> 
> I suspect most applications that use this driver aren't continuously re-opening database files so the problem doesn't manifests because the handle leak only happens when the application is shut down, in which case the handle is immediately cleaned up by the OS.

In my case, I'm doing it only once, on application startup. We verify the database, and if it's invalid, we want to rename the existing file (for reference), and create a new one. This issue arises because when using a provider that supports encryption, the call to ExecuteNonQuery internally, locks the file and throws an exception, before the external handle is available to the caller, to close things out. 

I've had to go with a customized version (see my note from October above) to handle our case. As noted above, it's not only just if it's not a valid DB file, even bad password attempts lock the file. I'm thinking that that's probably not intentional.

**Comment by JohnyL** [comment]
@ericnev Take a look at my code - I merely create a simple table, and procedure exits. Nothing fancy - no encryption, no passwords. As easy as it can be.

**Comment by ericnev** [comment]
> [@ericnev](https://github.com/ericnev) Take a look at my code - I merely create a simple table, and procedure exits. Nothing fancy - no encryption, no passwords. As easy as it can be.

Right. I'm wondering if you're running into an instance where the connection is pooled and thus still open in your case?

In blocks where I want to assure the file is not in use after, my finally looks like this:

        finally
        {
            if (connection.State == ConnectionState.Open)
                connection.Close();
            SqliteConnection.ClearPool(connection);
        }

**Comment by JohnyL** [comment]
@ericnev No, I'm not using pool. I was going to create it once and use it during app lifetime. Is it possible to track down what has been changed in those two versions that created this issue?

**Comment by cincuranet** [comment]
> Will it be fixed?

Yes. This issue is on my radar.

**Comment by JohnyL** [comment]
@cincuranet Thanks a lot!

**Comment by JohnyL** [comment]
Any updates on this? 😒

**Comment by cincuranet** [comment]
Same as https://github.com/dotnet/efcore/issues/35010#issuecomment-2576111943.
